lib/data/file/backend.js

```
@@ -0,0 +1,111 @@
+            request.on('data', data => {
+                // Disable data events as we need to wait for fs.write callback
+                request.pause();
+                hash.update(data);
+                return fs.write(fd, data, 0, data.length,
+                    err => {
+                        if (err) {
+                            log.error('error writing data', { error: err });
+                            return callback(errors.InternalError);
+                        }
+                        request.resume(); // Allow data events again
+                    });
+            });
+            request.on('error', () => {
+                log.error('error streaming data from request', { error: err });
+                return callback(errors.InternalError);
+            });
+            request.on('end', () => {
+                request.calculatedHash = hash.digest('hex');
+                log.debug('finished writing data', { key,
+                    calculatedHash: request.calculatedHash });
+                fs.close(fd);
+                return callback(null, key);
+            });
+        });
+    },
+
+    get: function getFile(key, range, reqUids, callback) {
+        const log = createLogger(reqUids);
+        const filePath = getFilePath(key);
+        log.debug('opening readStream to get data', { filePath });
+        const readStreamOptions = {
+            flags: 'r',
+            encoding: null,
+            fd: null,
+            autoClose: true,
+        };
+        if (range) {
+            readStreamOptions.start = range[0];
+            readStreamOptions.end = range[1];
+        }
+        const rs = fs.createReadStream(filePath, readStreamOptions);
+        return callback(null, rs);
+    },
+
+    delete: function delFile(key, reqUids, callback) {
+        const log = createLogger(reqUids);
+        const filePath = getFilePath(key);
+        log.debug('deleting file', { filePath });
+        return fs.unlink(filePath, callback);
```

`fs.unlink()` will give an `ENOENT` error to the callback so I think the error is not silently ignored. In any case, we should be aware if the metadata backend thinks an object still exists while it's missing on disk: it meands there is either a bug in some of our code or data/fs integrity issue - and we should be aware if that's the case.